toolchain/check/scope_stack.h

```diff
@@ -57,17 +46,22 @@ class ScopeStack {
   // Information about a scope that has been temporarily removed from the stack.
   struct SuspendedScope;
 
-  // Pushes a scope onto scope_stack_. NameScopeId::None is used for new
-  // scopes. lexical_lookup_has_load_error is used to limit diagnostics when a
-  // given namespace may contain a mix of both successful and failed name
-  // imports.
-  auto Push(SemIR::InstId scope_inst_id = SemIR::InstId::None,
-            SemIR::NameScopeId scope_id = SemIR::NameScopeId::None,
-            SemIR::SpecificId specific_id = SemIR::SpecificId::None,
-            bool lexical_lookup_has_load_error = false) -> void;
-
-  // Pops the top scope from scope_stack_, cleaning up names from
-  // lexical_lookup_.
+  // Pushes a scope for a declaration name's parameters.
+  auto PushForDeclName() -> void;
+
+  // Pushes an entity scope. Note functions must use `PushForFunction` instead.
```

Comment:
Optional:

```suggestion
  // Pushes a non-function entity scope. Functions must use `PushForFunction` instead.
```

Where practical, I prefer to flag exceptions to a general API contract (or at least flag that there _are_ exceptions) when first stating the contract.